### ISSUE TITLE
TIFF support

### ISSUE THREAD

**Issue by paperboyo** [guardian/grid#2395]
This issue is a (very) loose checklist for TIFF image support. Feel free to add/modify/delete/flush down the drain.

- [ ] new MIME type
- [ ] since our libraries should themselves deal with different flavours of TIFF correctly, we can support most provided we channel [transparent ones](https://github.com/guardian/grid/blob/master/image-loader/app/lib/imaging/FileMetadataReader.scala#L171) through current PNG route and flat ones through current JPEG route
- [ ] [adding](https://github.com/guardian/grid/blob/master/image-loader/app/lib/imaging/FileMetadataReader.scala#L13) [TIFF directory](https://github.com/drewnoakes/metadata-extractor/tree/master/Source/com/drew/metadata/tiff) to metadata
- [ ] collecting [colourModel](https://github.com/guardian/grid/blob/master/common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala#L146) information ([bit depth](https://github.com/guardian/grid/blob/master/image-loader/app/lib/imaging/FileMetadataReader.scala#L176), hasAlpha (transparency, really) etc.
- [ ] possibly collecting extra information without explicit control over these elements (just info about them): layers, paths etc
- [ ] possibly creating [prerenders](https://github.com/guardian/grid/blob/master/image-loader/app/model/ImageUpload.scala#L37) so that they can load faster _via_ imgOps

As some of the above operations are currently format-specific, adding TIFF provide a good opportunity to generalise them and bring some order to the galaxy.

**Comment by mbarton** [comment]
For reference, the work done to add PNG support:

https://github.com/guardian/grid/pull/1848
https://github.com/guardian/grid/pull/1912 (if we wanted to optimise the tiffs for consumption)
https://github.com/guardian/grid/pull/1959

**Comment by paperboyo** [comment]
Instead of my monumental #1959, I would add:
https://github.com/guardian/grid/pull/2375 (transparency sniffing leaky, follows https://github.com/guardian/grid/pull/1876)
https://github.com/guardian/grid/pull/1878 (transparency route)
https://github.com/guardian/grid/pull/1866 (no transparency > JPEG route)
https://github.com/guardian/grid/pull/1857 (format-specific metadata)

**Comment by mbarton** [comment]
https://github.com/guardian/grid/pull/2568

**Comment by paperboyo** [comment]
>  since our libraries should themselves deal with different flavours of TIFF correctly

This was/is a tad optimistic on my part. Some problems to think over/deal with:
- GraphicsMagick cannot correctly deal with LAB and CMYK TIFF files. ImageMagick can.
- there is a serious problem when dealing with layered TIFFs that contain transparency: if they do not have transparency saved in the composite layer¹, we would need to merge the individual layers ourselves²

¹ most unfortunately, and stupidly, a [Photoshop default](https://feedback.photoshop.com/photoshop_family/topics/default-tiff-save-options-to-be-save-transparency-on)

² non-simple layers (eg. adjustment layers) cannot be correctly merged using ImageMagick (or whatever really, like eg. libvips), we may be able to tell if the TIFF contains those problematic layers [via Exiftool](http://u88.n24.queensu.ca/exiftool/forum/index.php/topic,6435.msg47812.html#msg47812), though, and disable support only for them

**Comment by paperboyo** [comment]
As a lot of the outstanding issues depend on improved transparency detection, here is my current understanding on how to be most robust (will help sort out cases described in the previous comment).

We should use IM’s `identify -format '%[opaque]'`. It returns a string of boolean(s), eg. `TrueFalseTrue` calculating if there is transparency (`False`) for every layer in the file. We need to count them for all file formats which support layers. I assume the first one *always* represents a composite “background” layer. For all JPEGs and PNGs and for flat TIFFs, there will be only one (eg. `True`), we [should write it](https://github.com/guardian/grid/blob/master/image-loader/app/lib/imaging/FileMetadataReader.scala#L179) to our `hasAlpha`. (`%[opaque]`:`True` means `hasAplha`:`False`).

To correctly sense transparency for layered files, we need to look past that, though. If the count is >1, we could record a new property `hasLayers` under our (now badly named) `fileMetadata/colourModelInformation`.

So:

- If there is only one layer – do what we do to PNGs and JPEGs.
- If there is more:
  - If the first layer returns `False`, treat image as transparent PNGs, use only that first layer (it’s a correct composite). Use it also to create a thumbnail.

  - If the first layer returns `True`, the proper crazy part comes in:
    - count the number of `True` for all layers above the first, if the number is ≥ 1, treat image as JPEG (`hasAplha`:`False`), use only first layer (it’s a correct composite). Use it also to create a thumbnail.
    - if the number is `0` we arrived at the problematic case described in the previous comment: we have a layered TIFF file, which should be transparent (`hasAplha`:`True`), but we do not have a composite layer with transparency (stupid Photoshop saving default), only a useless one with white background.

We have two options:
– disallow uploading these files, warning users that they should check Save transparency
– look in metadata (² above) and attempt merging all the layers but the first one ourselves if we know that there are no “fancy” layers present that would result in the wrong merge result. If there are “fancy” layers – disallow uploading these files (and ask to resave).

Phew! I hope it makes sense. (Does it, though?)

**Comment by paperboyo** [comment]
TL;DR we may just disallow fancy layers when trasparency is not present in the composite, otherwise, this gets crazier… TBH, Adobe’s fault.

Looking at ² above, there may not be an easy way of making sure the file contains only the simple layers that ImageMagick would be able to merge correctly. Neither [Exiftool](https://www.sno.phy.queensu.ca/~phil/exiftool/TagNames/Photoshop.html), nor metadata-extractor contain readily useful info here as far as I can see (or rather: Adobe doesn’t provide it). One would need to dump all Photoshop metadata via Exiftool and grep for the presence of [any of those tags](https://www.adobe.com/devnet-apps/photoshop/fileformatashtml/#50577409_71546), then, any of ([layer effects](https://www.adobe.com/devnet-apps/photoshop/fileformatashtml/#50577409_22203) and, probably, more.

Here is a part of results for `exiftool -u -v1` command on a TIFF file with a Hue/Saturation Adjustment Layer (which ImageMagick would just ignore when merging):
```
  | | + [Layers directory with 2 entries, 4273176 bytes]
  | | | LayerCount = 2
  | | | + [Layer 1 of 2]
  | | | | LayerRectangles = 0 0 1115 1500
  | | | | LayerBlendModes = mron
  | | | | LayerOpacities = 255
  | | | | LayerNames = Layer 0
  | | | | LayerUnicodeNames = .Layer 0
  | | | | Photoshop_Layers_lnsr = ryal
  | | | | LayerIDs = 3
  | | | | Photoshop_Layers_clbl = .
  | | | | Photoshop_Layers_infx = 
  | | | | Photoshop_Layers_knko = 
  | | | | Photoshop_Layers_lspf = 
  | | | | Photoshop_Layers_lclr = 
  | | | | LayerModifyDates = .MIB8tsuc4...metadata..layerTimebuod..~..B.A
  | | | | Photoshop_Layers_fxrp = 
  | | | + [Layer 2 of 2]
  | | | | LayerRectangles = 0 0 0 0
  | | | | LayerBlendModes = mron
  | | | | LayerOpacities = 255
  | | | | LayerNames = Hue/Saturation 1
  | | | | Photoshop_Layers_hue2 = 
  | | | | Photoshop_Layers_hue2 = ..a..;.Y..-.-KiKi..............;.Y.
  | | | | LayerUnicodeNames = .Hue/Saturation 1
  | | | | Photoshop_Layers_lnsr = tnoc
  | | | | LayerIDs = 4
  | | | | Photoshop_Layers_clbl = .
  | | | | Photoshop_Layers_infx = 
  | | | | Photoshop_Layers_knko = 
  | | | | Photoshop_Layers_lspf = 
  | | | | Photoshop_Layers_lclr = 
  | | | | LayerModifyDates = .MIB8tsuc4...metadata..layerTimebuod%.~..B.A
  | | | | Photoshop_Layers_fxrp = 
```

Notice `Photoshop_Layers_hue2` entry indicating presence of a fancy layer that ImageMagick would just ignore on merge.

**Comment by paperboyo** [comment]
Save transparency option is recorded in metadata. Exiftool:
```
  | + [Photoshop Document Data directory, 475128 bytes]
  | | Photoshop_DocumentData_MTrn = 
  | | - Tag 'MTrn' (0 bytes)
```
`Mtrn`, `Mt16`, `Mt32`

[Described here:](https://www.adobe.com/devnet-apps/photoshop/fileformatashtml/#50577409_71546)

>  Saving Merged Transparency
> Key is 'Mtrn', 'Mt16' or 'Mt32' . There is no data associated with these keys.

**Comment by paperboyo** [comment]
Just in case we would want to support PSDs too (should be easy after sorting out the above). Maximise compatibility option ensures a merged result is saved to the first `[0]` layer (otherwise, this layer consists of gibberish/warning). There is a way to tell if the option has been used via [Image resource block](https://www.adobe.com/devnet-apps/photoshop/fileformatashtml/#50577409_34945) 0x0421 | 1057 (exiftool’s `HasRealMergedData = 1`).
